src/v2/controllers/vdeployment_controller/exec/reconciler.rs

```diff
@@ -0,0 +1,663 @@
+{
+    match state.reconcile_step {
+        VDeploymentReconcileStep::Done => true,
+        _ => false,
+    }
+}
+
+pub fn reconcile_error(state: &VDeploymentReconcileState) -> (res: bool)
+    ensures res == model_reconciler::reconcile_error(state@),
+{
+    match state.reconcile_step {
+        VDeploymentReconcileStep::Error => true,
+        _ => false,
+    }
+}
+
+// Features to be supported:
+// 1. rolling is supported, recreate is not. It waits for all old pods to stop running before adding new vrs.
+// 2. maxSurge and maxUnavailable are not supported.
+// 3. clean up old vrs is not supported. It's defined by "d.Spec.RevisionHistoryLimit" to keep that many replicas,
+//    which also requires the support for revisions inside annotations to sort old rs and status of vd (will not clean up if vd is paused, etc.)
+// 4. rollback is not supported.
+// *. When having multiple new vrs, k8s deterministically chooses the oldest one.
+//    We may not need to support this if we can prove this doesn't happen for our controller.
+pub fn reconcile_core(vd: &VDeployment, resp_o: Option<Response<VoidEResp>>, state: VDeploymentReconcileState) -> (res: (VDeploymentReconcileState, Option<Request<VoidEReq>>))
+    requires vd@.well_formed(),
+    ensures (res.0@, res.1.deep_view()) == model_reconciler::reconcile_core(vd@, resp_o.deep_view(), state@),
+{
+    let namespace = vd.metadata().namespace().unwrap();
+    match state.reconcile_step {
+        VDeploymentReconcileStep::Init => {
+            let req = KubeAPIRequest::ListRequest(KubeListRequest {
+                api_resource: VReplicaSet::api_resource(),
+                namespace: namespace,
+            });
+            let old_vrs_list = Vec::<VReplicaSet>::new();
+            // why this assertion is required
+            assert(old_vrs_list@.map_values(|vrs: VReplicaSet| vrs@) == Seq::<VReplicaSetView>::empty());
+            let state_prime = VDeploymentReconcileState {
+                reconcile_step: VDeploymentReconcileStep::AfterListVRS,
+                new_vrs: None,
+                old_vrs_list: old_vrs_list,
+            };
+            return (state_prime, Some(Request::KRequest(req)))
+        },
+        VDeploymentReconcileStep::AfterListVRS => {
+            if !(is_some_k_list_resp!(resp_o) && extract_some_k_list_resp_as_ref!(resp_o).is_ok()) {
+                return (error_state(state), None);
+            }
+            let objs = extract_some_k_list_resp!(resp_o).unwrap();
+            let vrs_list_or_none = objects_to_vrs_list(objs);
+            if vrs_list_or_none.is_none() {
+                return (error_state(state), None);
+            }
+            let (new_vrs_list, old_vrs_list) = filter_old_and_new_vrs(filter_vrs_list(vrs_list_or_none.clone().unwrap(), vd), vd);
+            // no .last().cloned() in verus because "The verifier does not yet support the following Rust feature: overloaded deref"
+            let state = VDeploymentReconcileState {
```

Comment:
Here you bind the variable `state` to a different value than the original argument